src/herder/TransactionQueue.cpp

```diff
@@ -0,0 +1,228 @@
+    {
+        if (m.empty())
+        {
+            continue;
+        }
+
+        std::set<std::shared_ptr<TxMap>> toRecalculate;
+
+        for (auto const& tx : dropTxs)
+        {
+            auto const& acc = tx->getSourceID();
+            auto const& txID = tx->getFullHash();
+            auto i = m.find(acc);
+            if (i != m.end())
+            {
+                auto& txs = i->second->mTransactions;
+                auto j = txs.find(txID);
+                if (j != txs.end())
+                {
+                    txs.erase(j);
+                    if (txs.empty())
+                    {
+                        m.erase(i);
+                    }
+                    else
+                    {
+                        toRecalculate.insert(i->second);
+                    }
+                }
+            }
+        }
+
+        for (auto txm : toRecalculate)
+        {
+            txm->recalculate();
+        }
+    }
+}
+
+bool
+TransactionQueue::contains(TransactionFramePtr tx) const
+{
+    return std::any_of(std::begin(mPendingTransactions),
+                       std::end(mPendingTransactions),
+                       [&](AccountTxMap const& map) {
+                           auto txMap = map.find(tx->getSourceID());
+                           if (txMap == map.end())
+                           {
+                               return false;
+                           }
+
+                           auto& transactions = txMap->second->mTransactions;
+                           return transactions.find(tx->getFullHash()) !=
+                                  std::end(transactions);
+                       });
+}
+
+AccountTransactionsState
+TransactionQueue::getState(AccountID const& accountID) const
+{
+    AccountTransactionsState state;
+
+    for (auto& map : mPendingTransactions)
+    {
+        auto i = map.find(accountID);
+        if (i != map.end())
+        {
+            auto& txmap = i->second;
+            state.mTotalFees += txmap->mCurrentState.mTotalFees;
+            state.mMaxSeq =
+                std::max(state.mMaxSeq, txmap->mCurrentState.mMaxSeq);
+        }
+    }
+
+    return state;
+}
+
+void
+TransactionQueue::shift()
+{
+    mPendingTransactions.erase(--mPendingTransactions.end());
+    mPendingTransactions.emplace_front();
+}
+
+int
+TransactionQueue::countPending(int index) const
+{
+    return countTxs(mPendingTransactions[index]);
+}
+
+std::shared_ptr<TxSetFrame>
+TransactionQueue::toTxSet(Hash lclHash) const
+{
+    auto result = std::make_shared<TxSetFrame>(lclHash);
+
+    for (auto const& m : mPendingTransactions)
+    {
+        for (auto const& pair : m)
+        {
+            for (auto const& tx : pair.second->mTransactions)
+            {
+                result->add(tx.second);
+            }
+        }
+    }
+
+    return result;
+}
+
+bool
+operator==(AccountTransactionsState const& x, AccountTransactionsState const& y)
+{
+    if (x.mMaxSeq != y.mMaxSeq)
```

Comment:
why not write this as `return !(x.mMaxSeq == x.mMaxSeq && x.mTotalFees == y.mTotalFees)` ?